Update documentation about how to cancel/kill invocations

tillrohrmann opened the pull request:
This commit updates the documentation about how to cancel/kill invocations
wrt to the latest changes. The main change is that invocations can now
be canceled via `DELETE <RESTATE_META_ENDPOINT>/invocations/<INVOCATION_IDENTIFIER>`.
Invocations can now be killed by adding the query parameter mode=kill:
`DELETE <RESTATE_META_ENDPOINT>/invocations/<INVOCATION_IDENTIFIER>?mode=kill`

This fixes https://github.com/restatedev/documentation/issues/233.

gvdongen commented on docs/references/errors.md:
```diff
@@ -7,7 +7,7 @@ slug: errors
 This page contains the list of error codes emitted by Restate components.
 
-## META0001
+## META0001 {#META0001}
```
Didn't it work without these tags? It should by using `/references/errors#meta0001`, I think?

tillrohrmann replied:
I guess it is generated by https://github.com/restatedev/documentation/blob/main/tools/generate_errors_page.sh#L31. Seems to be there to generate anchors that are capital letters.

gvdongen commented on docs/references/errors.md:
```diff
@@ -51,57 +51,58 @@ message GreetingRequest {
+## META0003 {#META0003}
 
-Cannot reach the endpoint to execute service discovery. Make sure:
+Cannot reach the deployment to execute service discovery. Make sure:
```
I think we should either use this term in the entire docs or stick to endpoint until we do the complete switch.
Also, if I read `deployment` the first thing I think is my Restate cluster deployment, but maybe that is just me.

gvdongen replied:
Especially, as long as we have this page called `deployment`: https://docs.restate.dev/restate/deployment

tillrohrmann replied:
This is something that was auto generated from the error descriptions. Maybe I can just remove the respective commit from the PR.

tillrohrmann replied:
I'll change it to "service deployment".

gvdongen commented on docs/references/errors.md:
```diff
@@ -111,32 +112,32 @@ The provided subscription is invalid. Subscriptions should have:
+## RT0003 {#RT0003}
 
-The invocation failed because the invoker received a message from a service endpoint larger than the `worker.invoker.message_size_limit`.
+The invocation failed because the invoker received a message from a service larger than the `worker.invoker.message_size_limit`.
```
```suggestion
The invocation failed because Restate received a message from a service larger than the `worker.invoker.message_size_limit`.
```

tillrohrmann replied:
Also auto generated. Let's fix the error messages here: https://github.com/restatedev/restate/tree/main/crates/errors/src/error_codes

gvdongen commented on docs/references/errors.md:
```diff
@@ -51,57 +51,58 @@ message GreetingRequest {
+Cannot propagate deployment/service metadata to Restate components. If you see this error when starting Restate, this might indicate a corrupted Meta storage.
 
-We recommend wiping the Meta storage and recreate it by registering endpoints in the same order they were registered before.
+We recommend wiping the Meta storage and recreate it by registering deployments in the same order they were registered before.
```
```suggestion
We recommend wiping the Meta storage and recreating it by registering deployments in the same order they were registered before.
```

tillrohrmann replied:
Will fix it in the restate repo.

gvdongen commented on docs/references/errors.md:
```diff
@@ -51,57 +51,58 @@ message GreetingRequest {
-* The provided `uri` is correct
-* The service endpoint is up and running
-* The Restate runtime can reach the service endpoint through the configured `uri`
+* The provided `uri`/`arn` is correct
```
```suggestion
* The provided `URI`/`ARN` is correct
```
Also for the other occurences of URI and ARN on this page.

tillrohrmann replied:
Will fix it in the restate repo.

gvdongen commented on docs/references/errors.md:
```diff
@@ -51,57 +51,58 @@ message GreetingRequest {
+## META0005 {#META0005}
 
-Cannot propagate endpoint/service metadata to Restate components. If you see this error when starting Restate, this might indicate a corrupted Meta storage.
+Cannot propagate deployment/service metadata to Restate components. If you see this error when starting Restate, this might indicate a corrupted Meta storage.
```
Is this one Restate component that tries to propagate that to another Restate component? It's not clear who is doing the 'propagation'.

tillrohrmann replied:
It is a meta error. Therefore, I believe it is the meta that fails to propagate the information to the rest of Restate.

gvdongen commented on docs/services/invocation.md:
```diff
@@ -141,38 +141,60 @@ The Invocation identifier is opaque and its current format should not be relied
+The response of the handler will then be propagated back to its caller where it will continue with the cancellation process.
 
+:::note
+Canceling an invocation is a non-blocking operation which in some rare cases needs to be retried by the user.
```
I don't really understand what is meant here.

tillrohrmann replied:
Cancellation is a non-blocking operation which means that it returns without the operation being complete. Sometimes cancellations have no effect which means that users have to trigger the cancellation again. If this is more helpful, then I'll update the part.

gvdongen commented on docs/services/invocation.md:
```diff
@@ -141,38 +141,60 @@ The Invocation identifier is opaque and its current format should not be relied
 ## Cancel an invocation
 
-:::caution
+If an invocation takes too long to complete and, therefore, is no longer of interest, you can cancel it.
```
```suggestion
If an invocation takes too long to complete or is no longer of interest, you can cancel it.
```
There can be many reasons for cancelling besides taking too long, I guess.

tillrohrmann replied:
True. Will change it.

gvdongen commented on docs/services/invocation.md:
```diff
@@ -141,38 +141,60 @@ The Invocation identifier is opaque and its current format should not be relied
+If an invocation takes too long to complete and, therefore, is no longer of interest, you can cancel it.
+Canceling an invocation allows it to free any resources it is holding and roll back any changes it has made so far.
+In order to roll back changes correctly, the service handlers need to contain the necessary compensation logic.
+If the required compensation logic is implemented, then the service state stays consistent even in the presence of cancellations.
```
I could interpret this as that the service state refers to Restate's K/V state, and that I need to compensate `ctx.set`.

gvdongen replied:
ooh okay, good to know

gvdongen commented on docs/services/invocation.md:
```diff
@@ -141,38 +141,60 @@ The Invocation identifier is opaque and its current format should not be relied
 
-At the moment, gracefully cancelling an invocation is not supported. It will be supported in future Restate releases.
+The cancellation process works recursively in the following way:
+First, Restate tries to cancel the leaves of the current invocation, i.e. complete pending sleeps and awakeables or try to cancel calls to other services.
```
```suggestion
First, Restate tries to cancel the leaves of the current invocation, i.e. interrupt ongoing sleeps and awakeables or try to cancel calls to other services.
```
The word `complete` probably only makes sense if you know that we use the term "completion" to complete a journal entry? I think interrupt might fit better here?

tillrohrmann replied:
True. Will change it.

gvdongen commented on docs/services/invocation.md:
```diff
@@ -141,38 +141,60 @@ The Invocation identifier is opaque and its current format should not be relied
 
-:::danger
+:::note
+Background calls and delayed calls will not be killed because they are considered detached from the originating call tree.
```
```suggestion
One-way calls and delayed calls will not be killed because they are considered detached from the originating call tree.
```

tillrohrmann replied:
Updated.

gvdongen commented on docs/services/invocation.md:
```diff
@@ -141,38 +141,60 @@ The Invocation identifier is opaque and its current format should not be relied
-A good example is when your code runs a non-deterministic action: If the invocation is suspended and re-scheduled afterwards, the replay of the invocation might lead to a different code path, generating an invalid journal and failing the invocation indefinitely.
-In such cases, you can request Restate to kill the invocation, thereby aborting its execution as soon as possible.
-If the invocation is ongoing, killing the invocation **will not** roll back its progress.
+Killing an invocation means that every call in the call tree of the invocation will be stopped immediately without given the service handler a chance to react.
```
```suggestion
Killing an invocation means that every call in the call tree of the invocation will be stopped immediately without giving the service handler a chance to react and execute compensation logic.
```

gvdongen commented on docs/services/invocation.md:
```diff
@@ -141,38 +141,60 @@ The Invocation identifier is opaque and its current format should not be relied
+If an invocation takes too long to complete and, therefore, is no longer of interest, you can cancel it.
+Canceling an invocation allows it to free any resources it is holding and roll back any changes it has made so far.
+In order to roll back changes correctly, the service handlers need to contain the necessary compensation logic.
+If the required compensation logic is implemented, then the service state stays consistent even in the presence of cancellations.
```
Do the docs explain somewhere how to implement compensation logic? Might be good to add otherwise?

gvdongen commented on docs/services/invocation.md:
```diff
@@ -141,38 +141,60 @@ The Invocation identifier is opaque and its current format should not be relied
-At the moment, gracefully cancelling an invocation is not supported. It will be supported in future Restate releases.
+The cancellation process works recursively in the following way:
+First, Restate tries to cancel the leaves of the current invocation, i.e. complete pending sleeps and awakeables or try to cancel calls to other services.
+Once the leaves are canceled, the service handler will be notified about the cancellation via a terminal error being thrown at the call site.
```
```suggestion
Once the leaves are canceled, a terminal error is thrown in the service handler at the point in the code the invocation had reached.
```

gvdongen reviewed (commented):
Hi Till, left some comments and thoughts. Let me know what you think.

tillrohrmann commented:
Thanks for the review @gvdongen. I've removed the commits that updated the error description as they should not be part of this PR. I've addressed most of your comments. PTAL.

gvdongen commented on docs/services/invocation.md:
````diff
@@ -141,38 +141,60 @@ The Invocation identifier is opaque and its current format should not be relied
 :::
 
+```shell
+$ curl -X DELETE <RESTATE_META_ENDPOINT>/invocations/<INVOCATION_IDENTIFIER>
````
```suggestion
$ curl -X DELETE <RESTATE_META_URL>/invocations/<INVOCATION_ID>
```

There is a PR open to unify this across the docs

tillrohrmann replied:
Good catch will update it accordingly.

gvdongen commented on docs/services/invocation.md:
````diff
@@ -141,38 +141,60 @@ The Invocation identifier is opaque and its current format should not be relied
 
 ```shell
-$ curl -X DELETE <RESTATE_META_ENDPOINT>/invocations/<INVOCATION_IDENTIFIER>
+$ curl -X DELETE <RESTATE_META_ENDPOINT>/invocations/<INVOCATION_IDENTIFIER>?mode=kill
````
```suggestion
$ curl -X DELETE <RESTATE_META_URL>/invocations/<INVOCATION_ID>?mode=kill
```

tillrohrmann replied:
Same here.

gvdongen reviewed (approved):
LGTM :+1: 

A few small suggestions for uniformity across the docs

tillrohrmann commented:
Thanks a lot for the review @gvdongen. Will merge this PR once we have created a new runtime release.
